Add code slicing feature

A-Ortner opened the pull request:
Title: Assemblé Code Slicing Feature

I added a new feature to Assemblé: Slice.
Before adding this feature, one had to manually select relevant cells one by one to submit them to an upstream repository. Now, the last relevant cell (e.g. output) can be selected and upon clicking on "slice" (new button in the toolbar) the program will automatically all dependent cells. During the process the code will be split up in a traversable program tree, then all relevant code lines are selected and the user receives an optimized, fully executable program that can be submitted to an upstream repository.

Changes: 
I had to import a library call python program analysis (https://github.com/andrewhead/python-program-analysis).
I then added a new widget to preview the sliced code and implemented the functionality in the index.ts file.

Testing:
So far, I only tested the feature manually with code examples.

A-Ortner committed:
added python program analysis as dependency
added new widget for slice preview

A-Ortner committed:
change tsconfig so python-program-analysis (ES2015) errors are avoided

A-Ortner committed:
added slicing functionality

micahjsmith committed:
Fix imports

micahjsmith committed:
Re apply noImplicitAny and noUnusedLocals

micahjsmith committed:
Add types to Loc class members

micahjsmith committed:
Remove noImplicitThis

A-Ortner committed:
merged with upstream/micahjsmith/dev-index

A-Ortner committed:
implemented the slicing functionality

A-Ortner committed:
bugfix: sliced lines are displayed in correct order

A-Ortner committed:
tslint prettier

A-Ortner committed:
added feature to submit sliced code to upstream repository

A-Ortner committed:
removed logs, added explanatory comments

micahjsmith commented:
Okay @A-Ortner I took a quick look and there is a lot of good code in there!

To help the reader of a PR (me in this case), it is best to provide some of the following:

* informative title of the PR
* motivation for the changes, i.e. what is the problem you are solving here
* summary of the changes, including screenshots if relevant
* test plan (how you tested the changes, including screenshots if relevant)

Can you please update the PR accordingly? Then I will review again afterwards once I have this necessary context

micahjsmith commented:
Btw I've been reviewing and will hope to complete the initial review tonight ET

micahjsmith reviewed (changes requested):
You've made a lot of great progress here and I was able to use the slice button pretty successfully, nice job

in addition to inline-comments, please address the following points:

* doc: add minimal documentation section to readme describing behavior of slice button
* doc: note that code will be collected even if it does not define a feature definition
* bug: behavior if the selected cell does not contain any code - user should expect an error message like "there is no code to slice"
* bug: need to exclude all non-code cells (such as markdown and raw). slice button does nothing if there exists a prior markdown cell that does not happen to also be valid python code, and uses the markdown cell if valid python code
*bug/limitation: code is not collected if out of order (i.e. cell 1 is a dependency of cell 2, but cell 2 dragged above cell 1). you can either fix this by collecting the content array using cell execution order (rather than cell layout order), or you can document this limitation

finally, can you please add some minimal unit tests for your code? especially getLocationFromCurrentCell should be easy to unit test. if the difficulty is still in configuring a unit test framework, then let me know and I will just set up for you so that you can just focus on writing the tests

Nice progress!

A-Ortner committed:
Added description of the code-slicing feature to readme

A-Ortner committed:
added check: if selected cell does not contain code, a warning is issued and the slice is not computed

A-Ortner committed:
added limitation: deleted cells will not be considered

A-Ortner commented on src/index.ts:
```diff
@@ -90,90 +186,147 @@ export class AssembleSubmitButtonExtension
+          return;
+        }
+
+        if (!(await isAuthenticated())) {
```
I explicitly put it there since I wanted the user to be able to have look at the slice without necessarily being authenticated. 
Can I leave it like that?

micahjsmith replied:
yes, that's a fair point. no need for user to authenticate until they confirm they want to submit. leave it like this and I will update later (to consolidate both buttons)

A-Ortner committed:
removed PreviewCodeWidget and used ConfirmWidget instead; added explanatory comment

A-Ortner committed:
tslint

A-Ortner committed:
moved some dependencies to devdependencies

A-Ortner commented:
Hi, 
thank you for reviewing my code!
I have added documentation and fixed some of the mentioned bugs.
I would be glad if you would help me set up a test framework. I have tried setting up jest but it does not work as of now.

Also, regarding this bug:
_need to exclude all non-code cells (such as markdown and raw). slice button does nothing if there exists a prior markdown cell that does not happen to also be valid python code, and uses the markdown cell if valid python code_
I was not quite aware of that, thanks for pointing it out. I am reading the documentation of JL, maybe there is a way to extract only code cells but no others.

A-Ortner committed:
refactoring, bugfix: markdown/raw cells are no longer considered for slicing

micahjsmith commented:
thanks for the changes, I will re-review and also give a shot at setting up a js unit test framework

A-Ortner committed:
added parameter checks to function getCodeFromSlice and getLocationFromCurrentCell

A-Ortner committed:
made getCodeFromSlice and getLocationFromCurrentCell global functions

A-Ortner commented:
Hi Micah, 
just checking in. Have you had time to have a look at my code yet?

> thanks for the changes, I will re-review and also give a shot at setting up a js unit test framework

micahjsmith commented on package.json:
```diff
@@ -42,22 +42,24 @@
     "@jupyterlab/settingregistry": "^2.0.1",
     "@lumino/disposable": "^1.3.1",
-    "react": "~16.9.0"
+    "cookiecutter": "^1.3.0",
```
still should not be in here, but I can delete after merge

micahjsmith reviewed (dismissed):
thanks for the changes. there are still some TODOs (which I will write down) but I think we should move forward with this PR.

to pass the lint step of CI, please run and commit

```
npm run tslint
```

then ping me when the tests pass and I will merge.

congrats on the work done so far, this is some cool functionality!

micahjsmith commented:
TODOs
1. strange behavior of slicing in some code cells

  for example, when the following cell is the very first cell in the notebook, when I try to slice it, the lines in the cell are reordered. lines within a cell should never be reordered.

before slicing
```
# some preliminaries...
import logging
from ballet.util.log import enable as enable_logger
enable_logger('ballet_predict_house_prices')
enable_logger('ballet')
import pandas as pd
pd.set_option('display.max_columns', None)
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
```

after slicing
```
sns.set()
import logging
from ballet.util.log import enable as enable_logger
enable_logger('ballet_predict_house_prices')
enable_logger('ballet')
import pandas as pd
pd.set_option('display.max_columns', None)
import matplotlib.pyplot as plt
import seaborn as sns
```

2. add unit tests
3. remove unused `cookiecutter` npm package

micahjsmith commented:
@A-Ortner I've added and configured [jest](https://jestjs.io/docs/getting-started) (https://github.com/ballet/ballet-assemble/commit/ef700f1ff10e33818cf64197973984eb8dfdbb52)

you can use it by adding a file `src/index.test.js` and creating test cases in
```
describe('index', () => {
  test('getCodeFromSlice does <something>', () => {
    // test case goes here
  });
});
```

A-Ortner commented:
Hi!
I fixed the bug (glad you found it, thanks for checking) and added the unittest. Running `npm run tslint` did not show me any tslint errors. I merged the new commits to the master branch (sorry for that confusion, I only realized that I created this pull request from development and not from master when it was too late). Should I create a new PR from the master branch?

You can see the changes here: https://github.com/A-Ortner/ballet-assemble

I still have problems with jest. I added all of the unit test I wanted to add, but I cannot run them because of the following error.
Some forums suggested to simply re-install ts-jest but that did not work and neither did more complex solutions.

```
Validation Error:

  Preset ts-jest/presets/js-with-babel is invalid:

  The "id" argument must be of type string. Received null
  TypeError [ERR_INVALID_ARG_TYPE]: The "id" argument must be of type string. Received null
    at validateString (internal/validators.js:124:11)
    at Module.require (internal/modules/cjs/loader.js:954:3)
    at require (internal/modules/cjs/helpers.js:92:18)
    at setupPreset (\\wsl$\Ubuntu-20.04\home\aort\development\ballet-assemble\node_modules\jest-config\build\normalize.js:230:14)
    at normalize (\\wsl$\Ubuntu-20.04\home\aort\development\ballet-assemble\node_modules\jest-config\build\normalize.js:554:15)
    at readConfig (\\wsl$\Ubuntu-20.04\home\aort\development\ballet-assemble\node_modules\jest-config\build\index.js:163:46)
    at readConfigs (\\wsl$\Ubuntu-20.04\home\aort\development\ballet-assemble\node_modules\jest-config\build\index.js:373:26)
    at \\wsl$\Ubuntu-20.04\home\aort\development\ballet-assemble\node_modules\@jest\core\build\cli\index.js:155:58
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (\\wsl$\Ubuntu-20.04\home\aort\development\ballet-assemble\node_modules\@jest\core\build\cli\index.js:108:24)

  Configuration Documentation:
  https://jestjs.io/docs/configuration.html


Process finished with exit code 1
```

micahjsmith committed:
Merge branch 'master' into development

micahjsmith commented:
Jumping back in to merge this helpful contribution :)

I will resolve any lint/test issues on master branch
